fvm/evm/emulator/state/base.go

```diff
@@ -0,0 +1,484 @@
+		if err != nil {
+			return err
+		}
+		v.accountSetupOnCommit = false
+
+	}
+
+	// if this is the first time we are setting up an
+	// code collection, store its collection id.
+	if v.codeSetupOnCommit {
+		err = v.ledger.SetValue(v.rootAddress[:], []byte(CodesStorageIDKey), v.codes.CollectionID())
+		if err != nil {
+			return err
+		}
+		v.codeSetupOnCommit = false
+	}
+	return nil
+}
+
+func (v *BaseView) fetchOrCreateCollection(path string) (collection *Collection, created bool, error error) {
+	collectionID, err := v.ledger.GetValue(v.rootAddress[:], []byte(path))
+	if err != nil {
+		return nil, false, err
+	}
+	if len(collectionID) == 0 {
+		collection, err = v.collectionProvider.NewCollection()
+		return collection, true, err
+	}
+	collection, err = v.collectionProvider.CollectionByID(collectionID)
+	return collection, false, err
+}
+
+func (v *BaseView) getAccount(addr gethCommon.Address) (*Account, error) {
+	// check cached accounts first
+	acc, found := v.cachedAccounts[addr]
+	if found {
+		return acc, nil
+	}
+
+	// then collect it from the account collection
+	data, err := v.accounts.Get(addr.Bytes())
+	if err != nil {
+		return nil, err
+	}
+	// decode it
+	acc, err = DecodeAccount(data)
+	if err != nil {
+		return nil, err
+	}
+	// cache it
+	if acc != nil {
+		v.cachedAccounts[addr] = acc
+	}
+	return acc, nil
+}
+
+func (v *BaseView) storeAccount(acc *Account) error {
+	data, err := acc.Encode()
+	if err != nil {
+		return err
+	}
+	// update the cache
+	v.cachedAccounts[acc.Address] = acc
+	return v.accounts.Set(acc.Address.Bytes(), data)
+}
+
+func (v *BaseView) getCode(addr gethCommon.Address) ([]byte, error) {
+	// check the cache first
+	code, found := v.cachedCodes[addr]
+	if found {
+		return code, nil
+	}
+	// check if account exist in cache and has codeHash
+	acc, found := v.cachedAccounts[addr]
+	if found && !acc.HasCode() {
+		return nil, nil
+	}
+	// then collect it from the code collection
+	code, err := v.codes.Get(addr.Bytes())
+	if err != nil {
+		return nil, err
+	}
+	if code != nil {
+		v.cachedCodes[addr] = code
+	}
+	return code, nil
+}
+
+func (v *BaseView) storeCode(addr gethCommon.Address, code []byte) error {
+	if len(code) == 0 {
+		delete(v.cachedCodes, addr)
+		return v.codes.Remove(addr.Bytes())
+	}
+	v.cachedCodes[addr] = code
+	return v.codes.Set(addr.Bytes(), code)
+}
+
+func (v *BaseView) getSlot(sk types.SlotAddress) (gethCommon.Hash, error) {
+	value, found := v.cachedSlots[sk]
+	if found {
+		return value, nil
+	}
+
+	col, err := v.getSlotCollection(sk.Address)
+	if err != nil {
+		return gethCommon.Hash{}, err
+	}
+
+	val, err := col.Get(sk.Key.Bytes())
+	if err != nil {
+		return gethCommon.Hash{}, err
+	}
+	value = gethCommon.BytesToHash(val)
+	v.cachedSlots[sk] = value
+	return value, nil
+}
+
+func (v *BaseView) storeSlot(sk types.SlotAddress, data gethCommon.Hash) error {
+	col, err := v.getSlotCollection(sk.Address)
+	if err != nil {
+		return err
+	}
+	v.cachedSlots[sk] = data
```

Comment:
Since `storeCode()` removes empty code from `v.cachedCodes`, do we also want to remove empty slot from `v.cachedSlots` here for consistency?

Reply:
fixed it, to stay consistent.